diff --git a/src/Extraction.cpp b/src/Extraction.cpp
new file mode 100644
index 0000000..173f88a
--- /dev/null
+++ b/src/Extraction.cpp
@@ -0,0 +1,434 @@
+#define _USE_MATH_DEFINES
+#include <cmath>
+#include <cstring>
+#include <fstream>
+#include <iostream>
+
+#include "convergence.hpp"  //How satisfying to see that it is in alphabetical order.
+#include "epot_bicgstabsolver.hpp"
+#include "epot_efield.hpp"
+#include "error.hpp"
+#include "func_solid.hpp"
+#include "geometry.hpp"
+#include "geomplotter.hpp"
+#include "ibsimu.hpp"
+#include "meshvectorfield.hpp"
+#include "particledatabase.hpp"
+#include "particlediagplotter.hpp"
+#include "particlestatistics.hpp"
+#include "random.hpp"
+
+#include </afs/ipp-garching.mpg.de/home/n/ndha/include/gperftools/profiler.h>
+
+//###########################################################################################################################################
+//######################### start of copy from the 026_Tabata_2000-CrosssectionsCpp folder ##################################################
+//###########################################################################################################################################
+
+const double ER = 13.605693009E-3; //This is the Rydberg energy in keV. NOTE: all the functions need keV as energy variable! Taken from the 2014 CODATA recommended values data set.
+
+double f1(double x, double c1, double c2) //Note the absence of the c0 prefactor. As a result, the cross sections are all in units of 1.0E-20 m2!
+{   //This is from the 'analytic expressions' section in the Tabata_2000 article
+    return c1*pow(x/ER, c2);
+}
+double f2(double x, double c1, double c2, double c3, double c4)
+{   //This is from the 'analytic expressions' section in the Tabata_2000 article
+    return f1(x, c1, c2)/( 1.0 + pow(x/c3, (c2+c4)) );
+}
+double f3(double x, double c1, double c2, double c3, double c4, double c5, double c6)
+{   //This is from the 'analytic expressions' section in the Tabata_2000 article
+    return f1(x, c1, c2)/( 1.0 + pow(x/c3, (c2+c4)) + pow(x/c5, (c2+c6)) );
+}
+double f4(double x, double c1, double c2, double c3, double c4, double c5, double c6, double c7, double c8)
+{   //This is from the 'analytic expressions' section in the Tabata_2000 article
+    return f1(x, c1, c2)*( 1.0 + pow(x/c3, (c4-c2)) )/( 1.0 + pow(x/c5, (c4+c6)) + pow(x/c7, (c4+c8)) );
+}
+
+double e01(double E1, double a1, double a2, double a3, double a4)
+{   //This is from the 'analytic expressions' section in the Tabata_2000 article
+    return f2(E1, a1, a2, a3, a4);
+}
+double e02(double E1, double a1, double a2, double a3, double a4, double a5, double a6)
+{   //This is from the 'analytic expressions' section in the Tabata_2000 article
+    return ( f2(E1, a1, a2, a3, a4) + a5*f2(E1/a6, a1, a2, a3, a4) ); //Note: because only the first four elements of at are used by the two f2 calls there is no need to pass a slice of the array.
+}
+double e06(double E1, double a1, double a2, double a3, double a4, double a5, double a6)
+{   //This is from the 'analytic expressions' section in the Tabata_2000 article
+    return f3(E1, a1, a2, a3, a4, a5, a6);
+}
+double e08(double E1, double a1, double a2, double a3, double a4, double a5, double a6, double a7, double a8, double a9)
+{   //This is from the 'analytic expressions' section in the Tabata_2000 article
+    return ( f2(E1, a1, a2, a3, a4) + f3(E1, a5, a2, a6, a7, a8, a9) );
+}
+double e11(double E1, double a1, double a2, double a3, double a4, double a5, double a6, double a7, double a8, double a9, double a10)
+{   //This is from the 'analytic expressions' section in the Tabata_2000 article
+    return ( f3(E1, a1, a2, a3, a4, a5, a6) + f2(E1, a7, a8, a9, a10) );
+}
+double e13(double E1, double a1, double a2, double a3, double a4, double a5, double a6, double a7, double a8, double a9, double a10, double a11, double a12)
+{   //This is from the 'analytic expressions' section in the Tabata_2000 article
+    return ( f3(E1, a1, a2, a3, a4, a5, a6) + f3(E1, a7, a8, a9, a10, a11, a12) );
+}
+
+double sigma_Hm_to_H0(double E)
+{   //This is reaction 47 from Tabata_2000
+    if( E < 2.37E-3 ) //Emin and Emax are given in Table I.
+      return 0.0;
+    if( E > 5.00E+1 )
+      return 0.0;
+    double Eth = 2.25E-3; //Eth in keV
+    double ci[10] = {4.19E-2, 1.89, 1.78E-1, -2.3E-1, 1.04, 8.7E-1, 1.65E+1, 1.088, 5.33E-3, 1.66E-1};
+    double E1 = E - Eth;
+    return e11(E1, ci[0], ci[1], ci[2], ci[3], ci[4], ci[5], ci[6], ci[7], ci[8], ci[9]);
+}
+double sigma_Hm_to_Hp(double E)
+{   //This is reaction 49 from Tabata_2000 (an Eth of 0.0 keV)
+    if( E < 1.00 ) //Emin and Emax are given in Table I.
+      return 0.0;
+    if( E > 3.00E+2 )
+      return 0.0;
+    double ci[6] = {1.75E-8, 3.88, 9.06E-1, -2.74E-1, 3.19, 1.19};
+    return e06(E, ci[0], ci[1], ci[2], ci[3], ci[4], ci[5]);
+}
+double sigma_H0_to_Hm(double E)
+{   //This is reaction 29 from Tabata_2000
+    if( E < 2.37E-2 ) //Emin and Emax are given in Table I.
+      return 0.0;
+    if( E > 9.11E+1 )
+      return 0.0;
+    double Eth = 2.1E-2; //Eth in keV
+    double ci[12] = {9.73E-3, 2.38, 1.39E-2, -5.51E-1, 7.7E-2, 2.12, 1.97E-6, 2.051, 5.5, 6.62E-1, 2.02E+1, 3.62}; //In C++11 these two lines could be done much more logical and 'pythonic'.
+    double E1 = E - Eth;
+    return e13(E1, ci[0], ci[1], ci[2], ci[3], ci[4], ci[5], ci[6], ci[7], ci[8], ci[9], ci[10], ci[11]);
+}
+double sigma_H0_to_Hp(double E)
+{   //This is reaction 31 from Tabata_2000
+    if( E < 5.62E-2 ) //Emin and Emax are given in Table I.
+      return 0.0;
+    if( E > 1.12E+2 )
+      return 0.0;
+    double Eth = 2.0E-2; //Eth in keV
+    double ci[6] = {2.53E-4, 1.728, 2.164, 7.74E-1, 1.639, 1.43E+1};
+    double E1 = E - Eth;
+    return e02(E1, ci[0], ci[1], ci[2], ci[3], ci[4], ci[5]);
+}
+double sigma_Hp_to_H0(double E)
+{   //This is reaction  6 from Tabata_2000
+    if( E < 3.16E-3 ) //Emin and Emax are given in Table I.
+      return 0.0;
+    if( E > 1.00E+2 )
+      return 0.0;
+    double Eth = 2.5E-3; //Eth in keV
+    double ci[9] = {2.12E+2, 1.721, 6.7E-4, 3.239E-1, 4.34E-3, 1.296, 1.42E-1, 9.34, 2.997};
+    double E1 = E - Eth;
+    return e08(E1, ci[0], ci[1], ci[2], ci[3], ci[4], ci[5], ci[6], ci[7], ci[8]);
+}
+
+//###########################################################################################################################################
+//######################### end of copy from the 026_Tabata_2000-CrosssectionsCpp folder ####################################################
+//###########################################################################################################################################
+
+class BGICallback : public TrajectoryHandlerCallback 
+{   //This is the first implementation of background gas interaction for the H- H0 H+ system, on basis of cross sections as function of energy. The MC approach is naive, just as a first step.
+    //Because the cross sections are just implemented as the computationally intensive fit functions from the Tabata_2000 article, this is slow. This will be improved by fitting e.g. polynomials.
+    //Or use interpolation? 
+    //Think about a proper way to implement the E/amu axis!! 
+    Geometry &_geom;
+    MTRandom  _randUniform;
+    MeshScalarField _bgneutr; 
+    double _sigmaHmH0, _sigmaHmHp, _sigmaH0Hm, _sigmaH0Hp, _sigmaHpH0, _sigmaHmal, _sigmaH0al, _brancHmH0, _brancHmHp, _brancH0Hm, _brancH0Hp;
+
+public:
+
+    BGICallback( Geometry &geom ) 
+        : _geom(geom), _randUniform(1), _bgneutr(geom) {
+
+        for( size_t i = 0; i < _bgneutr.size(0); i++ ) { //CURRENTLY ONE DENSITY IS MANUALLY SPECIFIED HERE 
+            for( size_t j = 0; j < _bgneutr.size(1); j++ ) {
+                for ( size_t k = 0; k < _bgneutr.size(2); k++ ) {                            
+                    _bgneutr( i, j, k ) =  1.0; //Density is in units of 1E+20 m-3             
+                }
+            }
+        }
+    }
+
+    virtual ~BGICallback() {}
+
+    virtual void operator()( ParticleBase *particle, ParticlePBase *x1, ParticlePBase *x2, ParticlePBase *xend, class ParticleDataBase *pdb )
+    {
+        Particle3D *partic = (Particle3D *) ( particle );
+        ParticleP3D *p3dx1 = (ParticleP3D *)( x1 );
+        ParticleP3D *p3dx2 = (ParticleP3D *)( x2 );
+        ParticleP3D *p3dxe = (ParticleP3D *)( xend );
+        ParticleDataBase3D *pdb3d = (ParticleDataBase3D *)( pdb );
+
+        Vec3D temploc1 = p3dx1->location();    
+        Vec3D temploc2 = p3dx2->location();  
+        Vec3D templodx = temploc2 - temploc1;     //Distance between the two points.
+        Vec3D temploxm = temploc1 + 0.5*templodx; //Middle point between the two locations.
+        Vec3D tempvexm = 0.5*(p3dx2->velocity() + p3dx1->velocity()); //Average velocity of the particle.
+        double tempEkeV = (0.5*MASS_U*tempvexm.ssqr()/CHARGE_E)*1.0E-3; //Energy at middle point in keV/amu, needed to calculate the rates. (.5mv^2/(m/u))*(1/e)*1.0E-3
+        double mcrand[0], psrand[0], mcphys, nsigma, distan, disnor; 
+        _randUniform.get( mcrand );
+
+        // std::cout << "Here comes some diagnostic information" << std::endl;
+        // std::cout << tempEkeV << std::endl;
+        // std::cout << sigma_Hm_to_H0(tempEkeV) << std::endl;
+        // std::cout << sigma_Hm_to_Hp(tempEkeV) << std::endl;
+        // std::cout << sigma_H0_to_Hm(tempEkeV) << std::endl;
+        // std::cout << sigma_H0_to_Hp(tempEkeV) << std::endl;
+        // std::cout << sigma_Hp_to_H0(tempEkeV) << std::endl;
+        // exit( 007 );
+     
+        if ( particle->q() == -CHARGE_E) //Charge exchange of negatively charged particles
+        {
+            double sigmHmH0 = sigma_Hm_to_H0(tempEkeV);
+            double sigmHmHp = sigma_Hm_to_Hp(tempEkeV);
+            double sigmHmal = sigmHmH0 + sigmHmHp;
+
+            nsigma = _bgneutr(temploxm) * sigmHmal;      //No need for sums over paths, the TrajectoryHandlerCallback is called every grid cell.
+            mcphys = exp( - templodx.norm2() * nsigma ); //Equation (2) from Stäbler_1994
+            if( mcrand[0] > mcphys ) //Because mcrand needs to be bigger, and mcphys is always positive, in principle it cannot be zero (which would give a pole error in the log function). 
+            {
+                distan = (-1.0/nsigma)*log(mcrand[0]);  //This is equation (3) from Stäbler_1994
+                disnor = distan/templodx.norm2();       //This should be a number between 0 and 1 indicating where the particle is between the two.
+                for ( size_t a = 0; a < p3dxe->size(); a ++ )
+                {
+                    p3dxe[0][a] = p3dx1[0][a] + disnor*(p3dx2[0][a] - p3dx1[0][a]); //The particle is created with linearly interpolated properties.
+                }
+                partic->set_status( PARTICLE_COLL ); //Also the end point needs to be updated
+
+                _randUniform.get( psrand ); //Determine which of the two processes is the one which is going to be the winner.
+                if( psrand[0] < (sigmHmH0/sigmHmal) )
+                {
+                    pdb3d->add_particle(particle->IQ(),  0.0, particle->m()/MASS_U, *p3dxe); //IQ, q, m, ParticleP3D   
+                }
+                else
+                {
+                    pdb3d->add_particle(particle->IQ(), +1.0, particle->m()/MASS_U, *p3dxe); //IQ, q, m, ParticleP3D   
+                }
+            }
+        }
+        else if ( particle->q() == 0.0) //Charge exchange of neutral particles
+        {   //There are two processes possible, H0 -> H- and H0 -> H+. This is the naive implementation, first I determine the chance at an event, then I share the events based on the sigma ratio.   
+            double sigmH0Hm = sigma_H0_to_Hm(tempEkeV);
+            double sigmH0Hp = sigma_H0_to_Hp(tempEkeV);
+            double sigmH0al = sigmH0Hm + sigmH0Hp;
+
+            nsigma = _bgneutr(temploxm) * sigmH0al;    //No need for sums over paths, the TrajectoryHandlerCallback is called every grid cell.
+            mcphys = exp( - templodx.norm2() * nsigma ); //Equation (2) from Stäbler_1994
+            if( mcrand[0] > mcphys ) //Because mcrand needs to be bigger, and mcphys is always positive, in principle it cannot be zero (which would give a pole error in the log function). 
+            {
+                distan = (-1.0/nsigma)*log(mcrand[0]);  //This is equation (3) from Stäbler_1994
+                disnor = distan/templodx.norm2();       //This should be a number between 0 and 1 indicating where the particle is between the two.
+                for ( size_t a = 0; a < p3dxe->size(); a ++ )
+                {
+                    p3dxe[0][a] = p3dx1[0][a] + disnor*(p3dx2[0][a] - p3dx1[0][a]); //The particle is created with linearly interpolated properties.
+                }
+                partic->set_status( PARTICLE_COLL ); //Also the end point needs to be updated
+
+                _randUniform.get( psrand ); //Determine which of the two processes is the one which is going to be the winner.
+                if( psrand[0] < (sigmH0Hm/sigmH0al) )
+                {
+                    pdb3d->add_particle(particle->IQ(), -1.0, particle->m()/MASS_U, *p3dxe); //IQ, q, m, ParticleP3D   
+                }
+                else
+                {
+                    pdb3d->add_particle(particle->IQ(), +1.0, particle->m()/MASS_U, *p3dxe); //IQ, q, m, ParticleP3D   
+                }
+            }
+        }
+        else if ( particle->q() == +CHARGE_E) //Charge exchange of positively charged particles
+        {
+            nsigma = _bgneutr(temploxm) * sigma_Hp_to_H0(tempEkeV);      //No need for sums over paths, the TrajectoryHandlerCallback is called every grid cell.
+            mcphys = exp( - templodx.norm2() * nsigma );  //Equation (2) from Stäbler_1994
+            if( mcrand[0] > mcphys ) //Because mcrand needs to be bigger, and mcphys is always positive, in principle it cannot be zero (which would give a pole error in the log function). 
+            {
+                distan = (-1.0/nsigma)*log(mcrand[0]);  //This is equation (3) from Stäbler_1994
+                disnor = distan/templodx.norm2();       //This should be a number between 0 and 1 indicating where the particle is between the two.
+                for ( size_t a = 0; a < p3dxe->size(); a ++ )
+                {
+                    p3dxe[0][a] = p3dx1[0][a] + disnor*(p3dx2[0][a] - p3dx1[0][a]); //The particle is created with linearly interpolated properties.
+                }
+                partic->set_status( PARTICLE_COLL ); //Also the end point needs to be updated
+                pdb3d->add_particle(particle->IQ(), 0.0, particle->m()/MASS_U, *p3dxe); //IQ, q, m, ParticleP3D   
+            }
+        }
+    }
+
+};
+
+
+void print_particle_statistics( ParticleStatistics stat ) //This function is shamelessly copied from the particledatabaseimp.hpp file. 
+{
+    std::cout << "Particles flown \t= " << stat.bound_collisions() << std::endl;
+    std::cout << "time limited \t\t= " << stat.end_time() <<  std::endl;
+    std::cout << "step count limited \t= " << stat.end_step() << std::endl;
+    std::cout << "bad definitions \t= " << stat.end_baddef() << std::endl;
+    for( size_t a = 1; a <= stat.number_of_boundaries(); a++ ) 
+    {
+        std::cout << "  beam to boundary " << a << " = " << stat.bound_current(a)
+                << " A (" << stat.bound_collisions(a) << " particles)" << std::endl;
+    }
+    std::cout << "total steps \t\t= " << stat.sum_steps() << std::endl;
+    std::cout << "avg steps per part(?) \t= " <<  stat.sum_steps()/(double)stat.bound_collisions() << std::endl;
+}
+
+void write_particle_statistics( ParticleStatistics stat, std::ofstream &myfile ) //This function writes particle statistics to an output file instead of the output stream
+{
+    myfile << "particles flown : \t" << stat.bound_collisions() << "\n";
+    myfile << "time lim: \t" << stat.end_time() <<  "\n";
+    myfile << "step lim: \t" << stat.end_step() << "\n";
+    myfile << "bad defs: \t" << stat.end_baddef() << "\n";
+    for( size_t a = 1; a <= stat.number_of_boundaries(); a++ ) 
+    {
+        myfile << "to boundary " << a << " [A]: \t" << stat.bound_current(a) << "\n";
+    }
+    for( size_t a = 1; a <= stat.number_of_boundaries(); a++ ) 
+    {
+        myfile << "to boundary " << a << " [ ]: \t" << stat.bound_collisions(a) << "\n";
+    }
+    myfile << "total steps: \t" << stat.sum_steps() << "\n";
+    myfile << "avg steps p: \t" <<  stat.sum_steps()/(double)stat.bound_collisions() << "\n";
+}
+
+void simu( void )
+{
+    std::string outpstr = "027_TrajectoryHandlerCallback_H_Naive3LevelSystem";
+    double diagframes[] = { 0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0, 1.1, 1.2, 1.3, 1.4, 1.5, 1.6, 1.7, 1.8, 1.9, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0 };
+    // double diagframes[] = { 0.0, 0.1 };
+
+    double h = 1.0e-2; //1 cm.
+    double sizereq[3] = {10.1, 10.0e-2, 10.0e-2 };
+    Vec3D origo( 0.0, -5.0e-2, -5.0e-2 );
+
+    uint32_t Ipart =  10000;   //Number of particles to be used.
+    double Imass = 1.0;        //Particle mass in atomic mass units
+    double Iener = 93.5E3;     //Energy of starting particles in eV
+
+    Int3D meshsize( (int)ceil(sizereq[0]/h)+1,
+                    (int)ceil(sizereq[1]/h)+1,
+                    (int)ceil(sizereq[2]/h)+1 );
+    Geometry geom( MODE_3D, meshsize, origo, h );
+
+    geom.set_boundary(  1, Bound(BOUND_DIRICHLET,    0.0  ) ); 
+    geom.set_boundary(  2, Bound(BOUND_NEUMANN,      0.0  ) ); 
+    geom.set_boundary(  3, Bound(BOUND_NEUMANN,      0.0  ) ); 
+    geom.set_boundary(  4, Bound(BOUND_NEUMANN,      0.0  ) ); 
+    geom.set_boundary(  5, Bound(BOUND_NEUMANN,      0.0  ) ); 
+    geom.set_boundary(  6, Bound(BOUND_NEUMANN,      0.0  ) ); 
+    geom.build_mesh();
+
+    GeomPlotter geomplotter( geom );
+    geomplotter.set_size(1920,1080);
+    geomplotter.set_font_size(20);
+    geomplotter.set_mesh( 1 );
+    geomplotter.plot_png(  outpstr + "geometry_empty.png"  );
+    geomplotter.plot_pdf(  outpstr + "geometry_empty.pdf"  );
+
+    bool fout[3] = { true, true, true }; 
+    MeshVectorField bfield( geom, fout );
+    // for( size_t i = 0; i < bfield.size(0); i++ ) { //If needed add a stationary field so that the difference between charged and neutral particles is clear.
+    //     for( size_t j = 0; j < bfield.size(1); j++ ) {
+    //         for ( size_t k = 0; k < bfield.size(2); k++ ) {                            
+    //             bfield.set( i, j, k, Vfc3D( 0, 0, -0.005 ) );
+    //         }
+    //     }
+    // }
+
+    EpotField epot( geom );
+    MeshScalarField sHcharge( geom );
+    EpotEfield efield( epot );
+    field_extrpl_e efldextrpl[6] = { FIELD_EXTRAPOLATE, FIELD_EXTRAPOLATE, 
+                                     FIELD_EXTRAPOLATE, FIELD_EXTRAPOLATE,
+                                     FIELD_EXTRAPOLATE, FIELD_EXTRAPOLATE };
+    efield.set_extrapolation( efldextrpl );
+
+    EpotBiCGSTABSolver solver( geom ); 
+
+    bool pmirror[6] = { false, false, false, false, false, false };
+    ParticleDataBase3D pdbH( geom ); //The hydrogen and electrons are kept in separate databases to facilitate diagnostics.
+    pdbH.set_mirror( pmirror );      
+
+    BGICallback bgicb( geom );
+    pdbH.set_trajectory_handler_callback( &bgicb );
+    // pdbH.set_save_all_points( true );  //All trajectory intersections with the mesh grid are saved if this is enabled. By default this is disabled: only the calculation points are saved.
+
+    ParticleDataBase3D pdbdiag( geom ); //This is a separate database to contain all the particles since one cannot assign two particle data bases to one plot.
+    pdbdiag.set_mirror( pmirror );
+
+    Emittance emit; //This is to monitor the convergence as the iterations are done.
+    Convergence conv;
+    conv.add_epot( epot );
+    conv.add_scharge( sHcharge );
+    conv.add_emittance( 0, emit );
+    
+    ParticleStatistics partistati( geom.number_of_boundaries() ); //This is to monitor the collisions with the boundaries. //uint32_t 10
+
+    solver.solve( epot, sHcharge );
+    efield.recalculate();
+
+    pdbH.clear();
+    double speed = sqrt( 2.0*Iener*CHARGE_E/(Imass*MASS_U) ); 
+    ParticleP3D p3d = ParticleP3D(0.0, 0.000, speed, 0.0, 0.0, 0.0, 0.0); //ParticleP3D (   double  t, double  x, double  vx, double  y, double  vy, double  z, double  vz )   emass
+    for( size_t i = 0; i < Ipart; i++ ) 
+    {
+        pdbH.add_particle(0.001, -1.0, 1.0, p3d); //double   IQ, double  q, double  m, const ParticleP3D &     x 
+    }
+
+    pdbH.iterate_trajectories( sHcharge, efield, bfield ); //NOTE: each function call to iterate_trajectories clears the space charge MeshScalarField, so two are necessary!!
+    
+    // pdbH.debug_print( std::cout );
+
+    geomplotter.set_size(1920,1080);
+    geomplotter.set_font_size(20);
+    geomplotter.set_epot( &epot );
+    geomplotter.set_scharge( &sHcharge );
+    geomplotter.set_particle_div( 1 );
+    geomplotter.set_particle_database( &pdbH );
+    geomplotter.set_view_si( VIEW_XY, 0.0 );
+    geomplotter.plot_png( outpstr + "geometry_iter_xy.png"  );
+    geomplotter.set_view_si( VIEW_XZ, 0.0 );
+    geomplotter.plot_png( outpstr + "geometry_iter_xz.png"  );
+
+    TrajectoryDiagnosticData tdata;
+    std::vector<trajectory_diagnostic_e> diagnostics;
+    diagnostics.push_back( DIAG_NO  );
+    diagnostics.push_back( DIAG_CHARGE );
+    diagnostics.push_back( DIAG_T );
+    diagnostics.push_back( DIAG_X );
+    diagnostics.push_back( DIAG_Y );
+    diagnostics.push_back( DIAG_Z ); 
+
+    char fname[50];
+    for ( size_t i = 0; i < sizeof(diagframes)/sizeof(diagframes[0]) ; i++ )
+    {
+        sprintf(fname, "Diagnostic_Data_%04.1f.txt", diagframes[i]);
+        pdbH.trajectories_at_plane( tdata, AXIS_X, diagframes[i], diagnostics ); //Save data
+        tdata.export_data( fname );
+    }
+
+    // geom.save( (outpstr + "_geometry.dat").c_str() ); //Save converged (?) data for a first comparison
+    // epot.save( (outpstr + "_epot.dat").c_str() );
+    // pdbH.save( (outpstr + "_H-_particle_database.dat").c_str() ); 
+    // pdbe.save( (outpstr + "_e-_particle_database.dat").c_str() ); 
+    // stotcharge.save( (outpstr + "_spacecharge.dat").c_str() );  
+
+}
+
+int main( int argc, char **argv )
+{
+    try 
+    {
+        ibsimu.set_message_threshold( MSG_VERBOSE, 1 );
+        ibsimu.set_thread_count( 4 );
+        ProfilerStart("Extraction_gperftools_profile.log");
+        simu();
+        ProfilerStop();
+    } 
+    catch ( Error e ) 
+    {
+        e.print_error_message( ibsimu.message( 0 ) );
+        exit( 1 );
+    }
+
+    return( 0 );
+}
\ No newline at end of file
diff --git a/src/id.hpp b/src/id.hpp
new file mode 100644
index 0000000..aa997c7
--- /dev/null
+++ b/src/id.hpp
@@ -0,0 +1 @@
+#define IBSIMU_GIT_ID "e8500a2, Mon Jul 25 15:47:35 2022 +0300"
